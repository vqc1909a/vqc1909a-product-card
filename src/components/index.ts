export { ProductCard }  from './ProductCard';
export { ProductTitle } from './ProductTitle';
export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
