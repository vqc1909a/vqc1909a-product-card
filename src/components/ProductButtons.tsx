import React from "react";
import { CSSProperties, useContext} from "react";
import { ProductCardContext } from './ProductCard';
import styles from "../styles/styles.module.css";

interface ProductButtonsProps {
  className?: string,
  style?: CSSProperties
}

export function ProductButtons({className, style}: ProductButtonsProps){
  const {counter, increaseBy, isMaxCountReached} = useContext(ProductCardContext);


  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style || {}}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>
      <div className={styles.countLabel}>{counter}</div>
      <button className={`${styles.buttonAdd} ${isMaxCountReached ? styles.disabled: ''}`} onClick={() => increaseBy(1)}>+</button>
    </div>
  )
}
