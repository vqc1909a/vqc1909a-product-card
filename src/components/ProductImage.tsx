import React from "react";
import { CSSProperties, useContext } from "react";
import { ProductCardContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from "../styles/styles.module.css";


interface ProductImageProps {
  img?: string,
  className?: string,
  style?: CSSProperties
}
export function ProductImage({img = "", className, style}: ProductImageProps ){
  let {product} = useContext(ProductCardContext);
  return (
    <img className={`${styles.productImg} ${className || ''}`} src={img || product?.img || noImage} alt={product?.title || 'no-image'} style={style || {}}/>
  )
}

