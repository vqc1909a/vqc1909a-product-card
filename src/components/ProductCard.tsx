import React from "react";
import { createContext, CSSProperties} from "react";
//!Tenemos que importar los estilos como modulos
import styles from "../styles/styles.module.css";
import useProduct from '../hooks/useProduct';
import { Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';
import {ProductImage} from "./ProductImage";
import {ProductButtons} from "./ProductButtons";
import {ProductTitle} from "./ProductTitle";

interface ProductCardProps {
  product: Product,
  // children?: ReactElement | ReactElement[]
  // children?: JSX.Element | JSX.Element[],
  children: (args: ProductCardHandlers) => JSX.Element,
  className?: string,
  style?: CSSProperties,
  onChange?: (args: onChangeArgs) => any,
  value?: number
  initialValues?: InitialValues 
}

interface ProductContext {
  product: Product,
  counter: number,
  increaseBy: (value: number) => void,
  maxCount?: number,
  isMaxCountReached?: boolean
}

export const ProductCardContext = createContext<ProductContext>({} as ProductContext);


export function ProductCard({children, product, className, style, onChange, value, initialValues}: ProductCardProps){
  const {counter, increaseBy, maxCount, isMaxCountReached, reset} = useProduct({onChange, product, value, initialValues});
  return (
    <ProductCardContext.Provider value={{product, counter, increaseBy, maxCount, isMaxCountReached}}>
      <div className={`${styles.productCard} ${className || ''}`} style={style || {}}>
        {/* El children es toda la funci√≥n, no te olvides de eso */}
        {/* De esta manera podemos exponer tantas propiedades como funciones de nuestro componente */}
        {children({
          count: counter,
          increaseBy,
          product,
          reset,
          maxCount,
          isMaxCountReached
        })}
      </div> 
    </ProductCardContext.Provider>
  )
}

// function ProductCard({product, children}: ProductCardProps) {
//   const {title, img} = product;
//   const {counter, increaseBy} = useProduct();

//   return (
//     <div className={styles.productCard}>
//       {children}
//       {/* <ProductImage img={img} title={title}></ProductImage>
//       <ProductTitle title={title}></ProductTitle>
//       <ProductButtons counter={counter} increaseBy={increaseBy} /> */}
//     </div>
//   )
// }

ProductCard.Image = ProductImage;
ProductCard.Title = ProductTitle;
ProductCard.Buttons = ProductButtons;

