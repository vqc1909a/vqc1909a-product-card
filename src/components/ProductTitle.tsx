import React from "react";
import { CSSProperties, useContext } from "react";
import { ProductCardContext } from './ProductCard';
import styles from "../styles/styles.module.css";

interface ProductTitleProps {
  title?: string,
  className?: string,
  style?: CSSProperties
}
export function ProductTitle({title = "", className, style}: ProductTitleProps){
  const {product} = useContext(ProductCardContext);
  const {title: titleProduct} = product;

  let titleToShow = title || titleProduct;

  return (
    <span className={`${styles.productDescription} ${className || ''}`} style={style || {}}>{titleToShow}</span>
  )
}

