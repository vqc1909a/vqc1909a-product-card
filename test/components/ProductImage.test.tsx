import React from "react";
import renderer from "react-test-renderer";
import { ProductImage } from '../../src/components/ProductImage';
import { ProductCard } from '../../src/components/ProductCard';

import { product2 } from '../data/products';


describe('ProductImage', () => {
    test('debe de mostrar el componente correctamente con la imagen personalizada', () => {
        const wrapper = renderer.create(
            <ProductImage img="https://www.adslzone.net/app/uploads-adslzone.net/2019/04/borrar-fondo-imagen.jpg"></ProductImage>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
    test('debe de mostrar iamgen del componente con la imagen del producto', () => { 
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage></ProductImage>
                    )
                }
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
})