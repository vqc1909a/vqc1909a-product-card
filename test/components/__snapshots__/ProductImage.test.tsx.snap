// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostrar el componente correctamente con la imagen personalizada 1`] = `
<img
  alt="no-image"
  className="productImg "
  src="https://www.adslzone.net/app/uploads-adslzone.net/2019/04/borrar-fondo-imagen.jpg"
  style={Object {}}
/>
`;

exports[`ProductImage debe de mostrar iamgen del componente con la imagen del producto 1`] = `
<div
  className="productCard "
  style={Object {}}
>
  <img
    alt="Coffee Mug - Con Imagen"
    className="productImg "
    src="/coffee-mug.png"
    style={Object {}}
  />
</div>
`;
